File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        
        Deque<int[]> onBridge = new ArrayDeque<>();
        
        int total_weight = 0;
        int time = 1;
        
        for(int truck : truck_weights){
            
            if(!onBridge.isEmpty() && onBridge.peek()[1] <= time){
                total_weight -= onBridge.poll()[0];
            }
            
            while(total_weight + truck > weight || onBridge.size() >= bridge_length){
                int out[] = onBridge.poll();
                total_weight -= out[0];
                time = out[1];
            }
            
            total_weight += truck;
            
            onBridge.add(new int[] {truck, time + bridge_length});
            time++;
            
        }
        
        return onBridge.peekLast()[1];
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    
    const onBridge = [];
    var index = 0;
    var time = 1;
    var total_weight = 0;
    var truckCnt = 0;
    
    for(let truck of truck_weights){
        
        if(onBridge.length && onBridge[index][1] <= time){
            total_weight -= onBridge[index][0];
            index++;
            truckCnt--;
        }
        
        while(truck + total_weight > weight || truckCnt >= bridge_length){
            total_weight -= onBridge[index][0];
            time = onBridge[index][1];
            index++;
            truckCnt--;
        }
        
        onBridge.push([truck, time+bridge_length]);
        total_weight += truck;
        truckCnt++;
        time++;
        
    }
    
    return onBridge.pop()[1];
}